src/jobserver.cc

```diff
@@ -0,0 +1,101 @@
+// Copyright 2024 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "jobserver.h"
+
+#include <cstring>
+#include <cassert>
+
+#include "util.h"
+
+bool Jobserver::Acquire() {
+  // The first token is implicitly handed to the ninja process, so don't
+  // acquire it from the jobserver
+  if (token_count_ == 0 || AcquireToken()) {
+    token_count_++;
+    return true;
+  }
+
+  return false;
+}
+
+void Jobserver::Release() {
+  assert(token_count_ >= 1);
+  token_count_--;
+
+  // Don't return first token to the jobserver, as it is implicitly handed
+  // to the ninja process
+  if (token_count_ > 0) {
+    ReleaseToken();
+  }
+}
+
```

Comment:
these functions are so small, the logic here might as well be combined with the implementation-specific function that they call, unless there is some reason to split the functions between private and public members of the class. If so please correct me.

If that happens, there's only one function left. Merging all the functions in jobserver.cc to the platform-specific files would also allow simplification of the parse function, making it smaller for each build-type and further reduce complexity.

Reply:
I put this logic in a common place as it is quite important (and a bit non-intuitive) to not acquire/release the first token. Having it duplicated in platform specific implementations increases the risk of getting it wrong.